File: recursion/fold.js
let {isEmpty, unprepend} = require('./lists');

function foldR(list, fn, initial) {
  if (isEmpty(list)) {
    return initial;
  }

  let [first, rest] = unprepend(list);

  return fn(first, foldR(rest, fn, initial));
}

function foldL(list, fn, initial) {
  if (isEmpty(list)) {
    return initial;
  }

  let [first, rest] = unprepend(list);

  return foldL(rest, fn, fn(initial, first));
}

let accumulateLeft = foldL;
let accumulateRight = foldR;

module.exports = {
  foldL,
  foldR,
  accumulateLeft,
  accumulateRight,
};

File: recursion/python/fold.py
from lists import is_empty, unprepend

def foldr(lst, fn, initial):
    if is_empty(lst):
        return initial

    first, rest = unprepend(lst)

    return fn(first, foldr(rest, fn, initial))

def foldl(lst, fn, initial):
    if is_empty(lst):
        return initial

    first, rest = unprepend(lst)

    return foldl(rest, fn, fn(initial, first))
